File: myLeetCode/Backtracking/46. Permutations.cpp
// 46. Permutations
// https://leetcode.com/problems/permutations/

#include "../printUtils.hpp"

class Solution {
public:
    std::vector<std::vector<int>> permute(std::vector<int>& nums) {
        std::vector<std::vector<int>> result;
        backtrack(result, nums, 0);
        return result;
    }

private:
    void backtrack(std::vector<std::vector<int>>& result, std::vector<int>& nums, int start) {
        if (start == nums.size()) {
            result.push_back(nums);
            return;
        }

        for (int i = start; i < nums.size(); ++i) {
            std::swap(nums[i], nums[start]);
            backtrack(result, nums, start + 1);
            std::swap(nums[i], nums[start]);
        }
    }
};

int main() {
    std::vector<int> nums = {1, 2, 3};
    printSet(Solution().permute(nums));

    return 0;
}

File: C++/代码随想录/Backtracking/46. Permutations.py
def permute(nums):
    def backtrack(start, end):
        if start == end:
            result.append(nums[:])
            return
        for i in range(start, end):
            nums[start], nums[i] = nums[i], nums[start]  # 交换
            backtrack(start + 1, end)
            nums[start], nums[i] = nums[i], nums[start]  # 恢复

    result = []
    backtrack(0, len(nums))
    return result

# 测试用例
nums = [1, 2, 3]
print(permute(nums))
